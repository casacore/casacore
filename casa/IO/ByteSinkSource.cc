//# SinkSource.cc: Class for read/write access to data in a given format
//# Copyright (C) 1996,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/aips.h>
#include <casa/IO/ByteSinkSource.h>


namespace casa { //# NAMESPACE CASA - BEGIN

ByteSinkSource::ByteSinkSource ()
{}

ByteSinkSource::ByteSinkSource (TypeIO* typeIO, Bool takeOver)
  : BaseSinkSource (typeIO, takeOver)
{}

ByteSinkSource::ByteSinkSource (const ByteSinkSource& sinkSource)
  : BaseSinkSource (sinkSource),
    ByteSink       (),
    ByteSource     ()
{}

ByteSinkSource& ByteSinkSource::operator= (const ByteSinkSource& sinkSource)
{
    BaseSinkSource::operator= (sinkSource);
    return *this;
}

ByteSinkSource::~ByteSinkSource()
{}

} //# NAMESPACE CASA - END

