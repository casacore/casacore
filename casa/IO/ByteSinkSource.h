//# ByteSinkSource.h: Class for read/write access to data in a given format
//# Copyright (C) 1996,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_BYTESINKSOURCE_H
#define CASA_BYTESINKSOURCE_H

#include <casa/aips.h>
#include <casa/IO/ByteSink.h>
#include <casa/IO/ByteSource.h>

namespace casa { //# NAMESPACE CASA - BEGIN

class TypeIO;

// <summary>Class for read/write access to data in a given format.</summary>

// <use visibility=export>

// <reviewed reviewer="Friso Olnon" date="1996/11/06" tests="tByteSinkSource" demos="">
// </reviewed>

// <prerequisite> 
//    <li> <linkto class=ByteSink>ByteSink</linkto> class
//    <li> <linkto class=ByteSource>ByteSource</linkto> class
//    <li> <linkto class=TypeIO>TypeIO</linkto> class and derived classes
// </prerequisite> 

// <etymology> 
// This class combines ByteSink and ByteSource.
// </etymology>

// <synopsis> 
// ByteSinkSource provides read/write access to a typed byte stream in the
// AIPS++ IO framework. It is derived from the classes <src>ByteSink</src>
// and <src>ByteSource</src>, so it combines their functionality.
// <p>
// The object is constructed using a typed byte stream. This stream
// is an instance of a class derived from class
// <linkto class=TypeIO>TypeIO</linkto>. This makes it possible to
// store the data in any format (e.g. CanonicalIO or RawIO).
// <br> In its turn TypeIO uses an instance of a class derived from class
// <linkto class=ByteIO>ByteIO</linkto>. This makes it possible to
// use any output stream (e.g. file, memory).
// </synopsis>

// <example>
// <srcblock>
// main 
// {
//     Bool valb = True;
//     RegularFileIO regularFileIO ("test.dat", ByteIO::New);
//     CanonicalIO canonicalIO(&regularFileIO);
//     ByteSinkSource  sinkSource(&canonicalIO);
//     sinkSource << valb;     // Write a boolean
//     sinkSource.seek (0);    // Reset to begin of IO stream
//     sinkSource >> valb;     // Read a boolean
//     cout << valb << endl;   // Print the boolean  
// }
// </srcblock>
// </example>

// <motivation> 
// This class makes it transparant to do IO with different devices and
// in different ways.
// </motivation>


class ByteSinkSource: public ByteSink, public ByteSource
{
public: 
    // Default constructor.
    // This creates an invalid object, but is present for convenience.
    ByteSinkSource();

    // Construct from given TypeIO object.  The constructor does not copy the
    // object, but only keeps a pointer to it. If takeOver is true the this
    // class will delete the supplied pointer. Otherwise the caller is
    // responsible for this.
    ByteSinkSource (TypeIO* typeIO, Bool takeOver=False);

    // The copy constructor uses reference semantics
    ByteSinkSource (const ByteSinkSource& sinkSource);

    // The assignment operator uses reference semantics
    ByteSinkSource& operator= (const ByteSinkSource& sinkSource);

    ~ByteSinkSource();
};



} //# NAMESPACE CASA - END

#endif
