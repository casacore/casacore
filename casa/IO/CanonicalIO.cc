//# CanonicalIO.cc: Class for IO in canonical format
//# Copyright (C) 1996,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/IO/CanonicalIO.h>
#include <casa/OS/CanonicalConversion.h>
#include <casa/IO/ByteIO.h>

namespace casa { //# NAMESPACE CASA - BEGIN

CanonicalIO::CanonicalIO (ByteIO* byteIO, uInt bufferLength, Bool takeOver)
: TypeIO          (byteIO, takeOver), 
  itsBuffer       (new char[bufferLength]),
  itsBufferLength (bufferLength)
{}

CanonicalIO::CanonicalIO (const CanonicalIO& that)
: TypeIO          (that), 
  itsBuffer       (new char[that.itsBufferLength]),
  itsBufferLength (that.itsBufferLength)
{}

CanonicalIO& CanonicalIO::operator= (const CanonicalIO& that)
{
    if (this != &that) {
        TypeIO::operator= (that);
	if (itsBufferLength != that.itsBufferLength) {
	    delete [] itsBuffer;
	    itsBufferLength = that.itsBufferLength;
	    itsBuffer = new char[itsBufferLength];
	}
    }
    return *this;
}

CanonicalIO::~CanonicalIO()
{
    delete [] itsBuffer;
}


uInt CanonicalIO::write (uInt nvalues, const Bool* value)
{
    return TypeIO::write (nvalues, value);
}

uInt CanonicalIO::write (uInt nvalues, const Char* value)
{
    if (CONVERT_CAN_CHAR) {
	if (nvalues * SIZE_CAN_CHAR <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_CHAR, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_CHAR];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_CHAR, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(Char), value);
    }
    return nvalues * SIZE_CAN_CHAR;
}

uInt CanonicalIO::write (uInt nvalues, const uChar* value)
{
    if (CONVERT_CAN_UCHAR) {
	if (nvalues * SIZE_CAN_UCHAR <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UCHAR, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_UCHAR];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UCHAR, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(uChar), value);
    }
    return nvalues * SIZE_CAN_UCHAR;
}

uInt CanonicalIO::write (uInt nvalues, const Short* value)
{
    if (CONVERT_CAN_SHORT) {
	if (nvalues * SIZE_CAN_SHORT <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_SHORT, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_SHORT];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_SHORT, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(Short), value);
    }
    return nvalues * SIZE_CAN_SHORT;
}

uInt CanonicalIO::write (uInt nvalues, const uShort* value)
{
    if (CONVERT_CAN_USHORT) {
	if (nvalues * SIZE_CAN_USHORT <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_USHORT, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_USHORT];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_USHORT, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(uShort), value);
    }
    return nvalues * SIZE_CAN_USHORT;
}

uInt CanonicalIO::write(uInt nvalues, const Int* value)
{
    if (CONVERT_CAN_INT) {
	if (nvalues * SIZE_CAN_INT <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_INT, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_INT];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_INT, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(Int), value);
    }
    return nvalues * SIZE_CAN_INT;
}

uInt CanonicalIO::write(uInt nvalues, const uInt* value)
{
    if (CONVERT_CAN_UINT) {
	if (nvalues * SIZE_CAN_UINT <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UINT, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_UINT];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UINT, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(uInt), value);
    }
    return nvalues * SIZE_CAN_UINT;
}

uInt CanonicalIO::write(uInt nvalues, const Int64* value)
{
    if (CONVERT_CAN_INT64) {
	if (nvalues * SIZE_CAN_INT64 <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_INT64, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_INT64];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_INT64, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(Int64), value);
    }
    return nvalues * SIZE_CAN_INT64;
}

uInt CanonicalIO::write(uInt nvalues, const uInt64* value)
{
    if (CONVERT_CAN_UINT64) {
	if (nvalues * SIZE_CAN_UINT64 <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UINT64, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_UINT64];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_UINT64, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(uInt64), value);
    }
    return nvalues * SIZE_CAN_UINT64;
}

uInt CanonicalIO::write(uInt nvalues, const float* value)
{
    if (CONVERT_CAN_FLOAT) {
	if (nvalues * SIZE_CAN_FLOAT <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_FLOAT, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_FLOAT];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_FLOAT, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(float), value);
    }
    return nvalues * SIZE_CAN_FLOAT;
}

uInt CanonicalIO::write(uInt nvalues, const double* value)
{
    if (CONVERT_CAN_DOUBLE) {
	if (nvalues * SIZE_CAN_DOUBLE <= itsBufferLength) {
	    CanonicalConversion::fromLocal (itsBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_DOUBLE, itsBuffer);
    	} else {
	    char* tempBuffer = new char [nvalues * SIZE_CAN_DOUBLE];
	    CanonicalConversion::fromLocal (tempBuffer, value, nvalues);
	    itsByteIO->write (nvalues * SIZE_CAN_DOUBLE, tempBuffer);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->write (nvalues * sizeof(double), value);
    }
    return nvalues * SIZE_CAN_DOUBLE;
}

uInt CanonicalIO::write (uInt nvalues, const Complex* value)
{
    return TypeIO::write (nvalues, value);
}

uInt CanonicalIO::write (uInt nvalues, const DComplex* value)
{
    return TypeIO::write (nvalues, value);
}

uInt CanonicalIO::write (uInt nvalues, const String* value)
{
    return TypeIO::write (nvalues, value);
}


uInt CanonicalIO::read (uInt nvalues, Bool* value)
{
    return TypeIO::read (nvalues, value);
}

uInt CanonicalIO::read (uInt nvalues, Char* value)
{
    if (CONVERT_CAN_CHAR) {
	if (nvalues * SIZE_CAN_CHAR <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_CHAR, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_CHAR];
	    itsByteIO->read (nvalues * SIZE_CAN_CHAR, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(Char), value);
    }
    return nvalues * SIZE_CAN_CHAR;
}

uInt CanonicalIO::read (uInt nvalues, uChar* value)
{
    if (CONVERT_CAN_UCHAR) {
	if (nvalues * SIZE_CAN_UCHAR <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_UCHAR, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_UCHAR];
	    itsByteIO->read (nvalues * SIZE_CAN_UCHAR, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(uChar), value);
    }
    return nvalues * SIZE_CAN_UCHAR;
}

uInt CanonicalIO::read (uInt nvalues, Short* value)
{
    if (CONVERT_CAN_SHORT) {
	if (nvalues * SIZE_CAN_SHORT <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_SHORT, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_SHORT];
	    itsByteIO->read (nvalues * SIZE_CAN_SHORT, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(Short), value);
    }
    return nvalues * SIZE_CAN_SHORT;
}

uInt CanonicalIO::read (uInt nvalues, uShort* value)
{
    if (CONVERT_CAN_USHORT) {
	if (nvalues * SIZE_CAN_USHORT <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_USHORT, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_USHORT];
	    itsByteIO->read (nvalues * SIZE_CAN_USHORT, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(uShort), value);
    }
    return nvalues * SIZE_CAN_USHORT;
}

uInt CanonicalIO::read (uInt nvalues, Int* value)
{
    if (CONVERT_CAN_INT) {
	if (nvalues * SIZE_CAN_INT <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_INT, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_INT];
	    itsByteIO->read (nvalues * SIZE_CAN_INT, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(Int), value);
    }
    return nvalues * SIZE_CAN_INT;
}

uInt CanonicalIO::read (uInt nvalues, uInt* value)
{
    if (CONVERT_CAN_UINT) {
	if (nvalues * SIZE_CAN_UINT <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_UINT, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_UINT];
	    itsByteIO->read (nvalues * SIZE_CAN_UINT, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(uInt), value);
    }
    return nvalues * SIZE_CAN_UINT;
}

uInt CanonicalIO::read (uInt nvalues, Int64* value)
{
    if (CONVERT_CAN_INT64) {
	if (nvalues * SIZE_CAN_INT64 <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_INT64, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_INT64];
	    itsByteIO->read (nvalues * SIZE_CAN_INT64, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(Int64), value);
    }
    return nvalues * SIZE_CAN_INT64;
}

uInt CanonicalIO::read (uInt nvalues, uInt64* value)
{
    if (CONVERT_CAN_UINT64) {
	if (nvalues * SIZE_CAN_UINT64 <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_UINT64, itsBuffer);
	    CanonicalConversion::toLocal(value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_UINT64];
	    itsByteIO->read (nvalues * SIZE_CAN_UINT64, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(uInt64), value);
    }
    return nvalues * SIZE_CAN_UINT64;
}

uInt CanonicalIO::read (uInt nvalues, float* value)
{
    if (CONVERT_CAN_FLOAT) {
	if (nvalues * SIZE_CAN_FLOAT <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_FLOAT, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_FLOAT];
	    itsByteIO->read (nvalues * SIZE_CAN_FLOAT, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(float), value);
    }
    return nvalues * SIZE_CAN_FLOAT;
}

uInt CanonicalIO::read (uInt nvalues, double* value)
{
    if (CONVERT_CAN_DOUBLE) {
	if (nvalues * SIZE_CAN_DOUBLE <= itsBufferLength) {
	    itsByteIO->read (nvalues * SIZE_CAN_DOUBLE, itsBuffer);
	    CanonicalConversion::toLocal (value, itsBuffer, nvalues);
	} else {
	    char* tempBuffer = new char[nvalues * SIZE_CAN_DOUBLE];
	    itsByteIO->read (nvalues * SIZE_CAN_DOUBLE, tempBuffer);
	    CanonicalConversion::toLocal (value, tempBuffer, nvalues);
	    delete [] tempBuffer;
	}
    } else {
	itsByteIO->read (nvalues * sizeof(double), value);
    }
    return nvalues * SIZE_CAN_DOUBLE;
}

uInt CanonicalIO::read (uInt nvalues, Complex* value)
{
    return TypeIO::read (nvalues, value);
}

uInt CanonicalIO::read (uInt nvalues, DComplex* value)
{
    return TypeIO::read (nvalues, value);
}

uInt CanonicalIO::read (uInt nvalues, String* value)
{
    return TypeIO::read (nvalues, value);
}

} //# NAMESPACE CASA - END

