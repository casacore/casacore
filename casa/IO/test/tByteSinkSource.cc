//# tByteSinkSource.cc: Test program for class ByteSinkSource
//# Copyright (C) 1996,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/aips.h>
#include <casa/iostream.h>
#include <casa/IO/ByteSinkSource.h>
#include <casa/IO/CanonicalIO.h>
#include <casa/IO/RegularFileIO.h>
#include <casa/OS/RegularFile.h>
#include <casa/OS/Path.h>
#include <casa/Exceptions.h>

#include <casa/namespace.h>
const Int nrOfTests = 1;

int main ()
{
    {
	RegularFileIO regularFileIO (Path("tByteSinkSource_tmp.dat"),
				     ByteIO::New);
    }
    {
	RegularFileIO regularFileIO (Path("tByteSinkSource_tmp.dat1"),
				     ByteIO::New);
	CanonicalIO canonicalIO(&regularFileIO);
	CanonicalIO canonicalIO1(canonicalIO);
	canonicalIO1 = canonicalIO;

	ByteSinkSource sinkSource(&canonicalIO);
	ByteSinkSource sinkSource1(sinkSource);
	sinkSource1 = sinkSource; 
	ByteSinkSource sinkSource2(&(sinkSource.typeIO()));
    }
    {
	Bool     testBool = True;
	Short    testShort = -30;
	uShort   testuShort = 10;
	Int      testInt = -20;
	uInt     testuInt = 80;
	Int64    testInt64 = -100000;
	uInt64   testuInt64 = 100000;
	Float    testFloat = 18.45;
	Double   testDouble = 23.987; 
	Complex  testComplex(2,3);
	DComplex testDComplex(2.5,3.8);
	Char     testChar = 'A';
	uChar    testuChar = 'B';
	String   testString("This is a teststring");

	RegularFileIO regularFileIO (Path("tByteSinkSource_tmp.dat1"),
				     ByteIO::New);
	CanonicalIO canonicalIO(&regularFileIO);
	ByteSinkSource  sinkSource(&canonicalIO);

	cout << sinkSource.isReadable() << endl;
	cout << sinkSource.isWritable() << endl;
	cout << sinkSource.isSeekable() << endl;

	sinkSource << testBool;
	sinkSource << testShort;
	sinkSource << testuShort;
	sinkSource << testInt;
	sinkSource << testuInt;
	sinkSource << testInt64;
	sinkSource << testuInt64;
	sinkSource << testFloat;
	sinkSource << testDouble;
	sinkSource << testComplex;
	sinkSource << testDComplex;
	sinkSource << testChar;
	sinkSource << testuChar;
	sinkSource << testString;

	sinkSource.write (nrOfTests, &testBool);
	sinkSource.write (nrOfTests, &testShort);
	sinkSource.write (nrOfTests, &testuShort);
	sinkSource.write (nrOfTests, &testInt);
	sinkSource.write (nrOfTests, &testuInt);
	sinkSource.write (nrOfTests, &testInt64);
	sinkSource.write (nrOfTests, &testuInt64);
	sinkSource.write (nrOfTests, &testFloat);
	sinkSource.write (nrOfTests, &testDouble);
	sinkSource.write (nrOfTests, &testComplex);
	sinkSource.write (nrOfTests, &testDComplex);
	sinkSource.write (nrOfTests, &testChar);
	sinkSource.write (nrOfTests, &testuChar);
	sinkSource.write (nrOfTests, &testString);

	sinkSource.seek (0);

	sinkSource >> testBool;
	sinkSource >> testShort;
	sinkSource >> testuShort;
	sinkSource >> testInt;
	sinkSource >> testuInt;
	sinkSource >> testInt64;
	sinkSource >> testuInt64;
	sinkSource >> testFloat;
	sinkSource >> testDouble;
	sinkSource >> testComplex;
	sinkSource >> testDComplex;
	sinkSource >> testChar;
	sinkSource >> testuChar;
	sinkSource >> testString;

	cout << testBool      << endl;
	cout << testShort     << endl;
	cout << testuShort    << endl;
	cout << testInt       << endl;
	cout << testuInt      << endl;
	cout << testInt64     << endl;
	cout << testuInt64    << endl;
	cout << testFloat     << endl;
	cout << testDouble    << endl;
	cout << testComplex   << endl;
	cout << testDComplex  << endl;
	cout << testChar      << endl;
	cout << testuChar     << endl;
	cout << testString    << endl;

	sinkSource.read (nrOfTests, &testBool);
	sinkSource.read (nrOfTests, &testShort);
	sinkSource.read (nrOfTests, &testuShort);
	sinkSource.read (nrOfTests, &testInt);
	sinkSource.read (nrOfTests, &testuInt);
	sinkSource.read (nrOfTests, &testInt64);
	sinkSource.read (nrOfTests, &testuInt64);
	sinkSource.read (nrOfTests, &testFloat);
	sinkSource.read (nrOfTests, &testDouble);
	sinkSource.read (nrOfTests, &testComplex);
	sinkSource.read (nrOfTests, &testDComplex);
	sinkSource.read (nrOfTests, &testChar);
	sinkSource.read (nrOfTests, &testuChar);
	sinkSource.read (nrOfTests, &testString);
	
	cout << testBool      << endl;
	cout << testShort     << endl;
	cout << testuShort    << endl;
	cout << testInt       << endl;
	cout << testuInt      << endl;
	cout << testInt64     << endl;
	cout << testuInt64    << endl;
	cout << testFloat     << endl;
	cout << testDouble    << endl;
	cout << testComplex   << endl;
	cout << testDComplex  << endl;
	cout << testChar      << endl;
	cout << testuChar     << endl;
	cout << testString    << endl;
    }
    {
	RegularFileIO regularFileIO (Path("tByteSinkSource_tmp.dat1"),
				     ByteIO::Update);
	CanonicalIO canonicalIO(&regularFileIO, 5);
	ByteSinkSource sinkSource(&canonicalIO);

	String   testString;
	sinkSource << "This is a teststring";
	sinkSource.seek (0);
	sinkSource >> testString;
	cout << testString    << endl;
    }
    {
	try {
	    RegularFileIO regularFileIO (Path("tByteSinkSource_tmp.dat"),
					 ByteIO::NewNoReplace);
	} 
	catch (AipsError x)  {
	    cout << x.getMesg () << endl;
	} 
    }
    return 0;
}
