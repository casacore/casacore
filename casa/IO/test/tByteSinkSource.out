1
1
1
1
-30
10
-20
80
-100000
100000
18.45
23.987
(2,3)
(2.5,3.8)
A
B
This is a teststring
1
-30
10
-20
80
-100000
100000
18.45
23.987
(2,3)
(2.5,3.8)
A
B
This is a teststring
This is a teststring
RegularFileIO: new file tByteSinkSource_tmp.dat already exists
